src/core/compat/reversediff.jl

```diff
@@ -81,15 +79,13 @@ end
         f::F
         x::Tx
     end
-    function Memoization._get!(f::Union{Function, Type}, d::IdDict, keys::Tuple{Tuple{RDTapeKey}, Any})
+    function Memoization._get!(f, d::IdDict, keys::Tuple{Tuple{RDTapeKey}, Any})
         key = keys[1][1]
-        return Memoization._get!(f, d, (typeof(key.f), typeof(key.x), size(key.x)))
+        return Memoization._get!(f, d, (key.f, typeof(key.x), size(key.x), Threads.threadid()))
```

Comment:
We need the type here otherwise the memoization will not work when the functions are different but not identical. I think we construct the closure in every `gradient_logp` call so the function differentiated is different every time but its type is the same. This change will actually break tape memoization.

Reply:
This change is taken from the other PR (on which the changes here are based), which tries to fix https://github.com/TuringLang/Turing.jl/issues/1393. Memoization leads to incorrect results if it is only based on the type here.

Reply:
Let me see.

Reply:
Memoization seems to work still: https://github.com/TuringLang/Turing.jl/pull/1414#issuecomment-698215278